src/OmniSharp.Roslyn.CSharp/Services/Navigation/GoToDefinitionHelpers.cs

```diff
@@ -0,0 +1,51 @@
+﻿#nullable enable
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.FindSymbols;
+using OmniSharp.Extensions;
+using OmniSharp.Options;
+using System.Threading.Tasks;
+
+namespace OmniSharp.Roslyn.CSharp.Services.Navigation
+{
+    internal static class GoToDefinitionHelpers
+    {
+        internal static async Task<ISymbol?> GetDefinitionSymbol(Document document, int line, int column)
+        {
+            var sourceText = await document.GetTextAsync();
+            var position = sourceText.GetPositionFromLineAndOffset(line, column);
+            var symbol = await SymbolFinder.FindSymbolAtPositionAsync(document, position);
+
+            return symbol switch
+            {
+                INamespaceSymbol => null,
+                // Always prefer the partial implementation over the definition
+                IMethodSymbol { IsPartialDefinition: true, PartialImplementationPart: var impl } => impl,
+                // Don't return property getters/settings/initers
+                IMethodSymbol { AssociatedSymbol: IPropertySymbol } => null,
+                _ => symbol
+            };
+        }
+
+        internal static async Task<FileLinePositionSpan?> GetMetadataMappedSpan(
+            Document document,
+            ISymbol symbol,
+            ExternalSourceServiceFactory externalSourceServiceFactory,
+            IExternalSourceService externalSourceService,
+            OmniSharpOptions options,
+            int timeout)
+        {
+
+            var cancellationToken = externalSourceServiceFactory.CreateCancellationToken(options, timeout);
```

Comment:
in the past the cancellation token was only used for the metadata stuff, but since you are refactoring into a cleaner v2 version anyway, I think it is time we start doing this properly and create the cancellation token based on the request timeout at the entry point to the handler and flow it through all the async APIs used. Would you mind changing this?